cloud/kyup/kyup.py

```diff
@@ -0,0 +1,357 @@
+
+def get_task_status(module, task_id):
+    req = '{"action":"getTask","authorization_key":"%s","data":{"task_id":%d}}'
+    count = retries
+    while count:
+        ret = api_request(module, req % (module.params.get('api_key'), int(task_id)) )
+        if 'container_id' in ret['data']['task']:
+            return ret['data']['task']['container_id']
+        count -= 1
+    module.fail_json(changed=False, msg = ret)
+
+def add_ssh_keys(module, container_id):
+    key_list = {}
+    ssh_keys = module.params['ssh_keys']
+    if ssh_keys is None or ssh_keys == '':
+        return
+    # cleanup any spaces put by the user
+    ssh_keys.replace(' ', '')
+
+    for i in ssh_keys.split(','):
+        key_list[i] = 0
+    # get the key IDs
+    req = '{"action":"%s","authorization_key":"%s","data":{%s}}'
+    ret = api_request(module, req % ('sshGetKeys', module.params.get('api_key'), ''))
+    for key in ret['data']['keys']:
+        if key['title'] in key_list:
+            key_list[key['title']] = key['key_id']
+    # add the keys to the container
+    key_req = '"key_id":%d,"container_id":%d'
+    for key in key_list:
+        if key_list[key] == 0:
+            module.fail_json(changed=True, msg = 'Unable to find key ID for key %s' % key)
+        api_request(module, req % ('sshInstallKey', module.params.get('api_key'), key_req % (int(key_list[key]), int(container_id) )))
+
+def create_container(module):
+    enc_key = module.params['enc_key'] or os.environ.get('KYUP_ENC_KEY', None)
+
+    if enc_key is None:
+        module.fail_json(changed=False, msg = 'ENC_KEY is required for creation of containers')
+
+    opt = { 'name' : None, 'password' : None, 'image' : None, 'dc_id' : None }
+    for i in opt:
+        opt[i] =  module.params.get(i)
+        if opt[i] is None:
+            module.fail_json(changed=False, msg = '%s parameter is required for creating a container' % i)
+
+    # check if the container already exists
+    req = '{"action":"cloudList","authorization_key":"%s","data":{}}'
+    ret = api_request(module, req % module.params.get('api_key'))
+    for container in ret['data']['list']:
+        if container['name'] == module.params.get('name'):
+            container['ip'] = container['ip'][0:-3]
+            module.exit_json(
+                changed=False,
+                ansible_facts = { 'kyup_container_name': container['name'], 'kyup_container_ip': container['ip'] },
+                msg = 'Container %s already exists' % module.params.get('name')
+            )
+
+    storage_type = module.params['storage_type']
+    if storage_type is None or storage_type == 'local':
+        storage_type = 0
+    elif storage_type == 'distributed':
+        storage_type = 1
+
+    container_id = 0
+    req = '{"action":"cloudCreate","authorization_key":"%s","data":{"name":"%s","password":"%s","image_name":"%s","datacenter_id":%d,"storage_type":%d,"resources":{"mem":%d,"hdd":%d,"cpu":%d,"bw":%d}}}'
```

Comment:
Please use the JSON serializer instead. Using string formatting to inject things into JSON could have unfortunate consequences.